### PR TITLE
CLN:F-string formatting

### PR DESCRIPTION
**Pull request by ShaharNaveh**
Ref to #29547 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/_libs/parsers.pyx [hunk 1]
```diff
@@ -741,8 +740,7 @@ cdef class TextReader:
                           self.parser.lines < hr):
                     msg = self.orig_header
                     if isinstance(msg, list):
-                        msg = "[%s], len of %d," % (
-                            ','.join(str(m) for m in msg), len(msg))
+                        msg = f"[{','.join(str(m) for m in msg)}], len of {len(msg)},"
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx]: what do you think about taking the .join outside of the fstring?

File: pandas/_libs/parsers.pyx [hunk 2]
```diff
@@ -684,9 +683,9 @@ cdef class TextReader:
             if ptr == NULL:
                 if not os.path.exists(source):
                     raise FileNotFoundError(
-                        ENOENT,
-                        'File {source} does not exist'.format(source=source),
-                        source)
+                            ENOENT,
+                            f'File {source} does not exist',
+                            source)
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx]: lint complaint

File: pandas/_libs/parsers.pyx [hunk 3]
```diff
@@ -768,10 +766,9 @@ cdef class TextReader:
 
                     if name == '':
                         if self.has_mi_columns:
-                            name = ('Unnamed: {i}_level_{lvl}'
-                                    .format(i=i, lvl=level))
+                            name = (f'Unnamed: {i}_level_{level}')
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx]: parens not needed

File: pandas/_libs/parsers.pyx [hunk 4]
```diff
@@ -845,10 +842,9 @@ cdef class TextReader:
 
             passed_count = len(header[0])
 
-            # if passed_count > field_count:
-            #     raise ParserError('Column names have %d fields, '
-            #                        'data has %d fields'
-            #                        % (passed_count, field_count))
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx]: is this related to the rest of the PR?  we definitely dont want commented-out code lying around, but it isnt obvious to me that enabling it is the right thing to do

File: pandas/_libs/tslib.pyx [hunk 5]
```diff
@@ -266,12 +266,7 @@ def format_array_from_datetime(ndarray[int64_t] values, object tz=None,
         elif basic_format:
 
             dt64_to_dtstruct(val, &dts)
-            res = '%d-%.2d-%.2d %.2d:%.2d:%.2d' % (dts.year,
-                                                   dts.month,
-                                                   dts.day,
-                                                   dts.hour,
-                                                   dts.min,
-                                                   dts.sec)
+            res = f'{dts:year-month.2d day.2d:hour.2d:min.2d:sec.2d}'
```
Comment (jbrockmendel) [inline pandas/_libs/tslib.pyx]: this looks wrong to me, but may just be a feature of fstrings im not away of.  id expect something more like `f"{dts.year}-{dts.month%2d}-{dts.day}..." but im not sure how to do the 2-digit formatting here

File: pandas/_libs/parsers.pyx (hunk 1 above)
Comment (MomIsBestFriend) [reply pandas/_libs/parsers.pyx]: I do think that this is better(taking the join out of the f-string), 
but how would you implement this?

by assigning another variable the values from the comprehension list? then just puting it in the f-string?

I'm not really confident about this at all.

Comment (jbrockmendel) [reply pandas/_libs/parsers.pyx]: yah, just on the previous line `joined = ','.join(str(m) for m in msg)`

> I'm not really confident about this at all.

It's really helpful that you're doing this, especially in the cython code which is harder to lint and still has some ancient idioms floating around.

File: pandas/_libs/parsers.pyx (hunk 4 above)
Comment (MomIsBestFriend) [reply pandas/_libs/parsers.pyx]: Gotta be honest, It just bothered me that when I run grep ```parsers.pyx``` is popping because of this.

I don't mind changing it back, if that really matters.

Comment (jbrockmendel) [reply pandas/_libs/parsers.pyx]: I completely understand.  If you grep for `# FIXME:` you'll find a bunch of "don't leave commented-out code" that I put in whenever I come across it and don't immediately know what to do with it.

Comment (MomIsBestFriend) [reply pandas/_libs/parsers.pyx]: so for the time being leaving it with the change?
or without?

Comment (jbrockmendel) [reply pandas/_libs/parsers.pyx]: pls revert and either a) add a FIXME comment or b) re-enable or delete in a separate PR

Comment (MomIsBestFriend) [reply pandas/_libs/parsers.pyx]: Will delete in a separate pr.
I think that this deserves a discussion on its own.

File: pandas/_libs/tslib.pyx [hunk 6]
```diff
@@ -266,20 +266,16 @@ def format_array_from_datetime(ndarray[int64_t] values, object tz=None,
...
 
             dt64_to_dtstruct(val, &dts)
-            res = '%d-%.2d-%.2d %.2d:%.2d:%.2d' % (dts.year,
-                                                   dts.month,
-                                                   dts.day,
-                                                   dts.hour,
-                                                   dts.min,
-                                                   dts.sec)
+            res = (f'{dts.year}-{dts.month}-{dts.day} '
+                   f'{dts.hour}-{dts.min}-{dts.sec}')
```
Comment (jbrockmendel) [inline pandas/_libs/tslib.pyx]: it looks like this is failing.  i suggest reverting this for now to get the tests passing and figure this out in a separate PR

Comment (WillAyd) [reply pandas/_libs/tslib.pyx]: I think just need zero padding with digits, so `f"{dts.year}-{dts.month:02}..."`

**Comment by MomIsBestFriend** [comment]
ping @jbrockmendel

File: pandas/_libs/tslib.pyx (hunk 6 above)
Comment (MomIsBestFriend) [reply pandas/_libs/tslib.pyx]: Will open a separate PR, and will try what @WillAyd suggested

File: pandas/_libs/parsers.pyx [hunk 7]
```diff
@@ -846,9 +844,9 @@ cdef class TextReader:
             passed_count = len(header[0])
 
             # if passed_count > field_count:
-            #     raise ParserError('Column names have %d fields, '
-            #                        'data has %d fields'
-            #                        % (passed_count, field_count))
+            #    raise ParserError('Column names have %d fields, '
+            #                      'data has %d fields'
+            #                      % (passed_count, field_count))
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx]: can you get this back to zero-diff.  otherwise the next person who comes along to review will go through the same song-and-dance about it

**Comment by jbrockmendel** [comment]
trailing whitespace

**Comment by MomIsBestFriend** [comment]
gentle desperate ping @jbrockmendel

**Comment by jbrockmendel** [comment]
LGTM cc @WillAyd

**Comment by jreback** [comment]
thanks @MomIsBestFriend 

definitely more readable with the f-strings, keep em coming!
